swupd/delta.go

```diff
@@ -42,152 +43,123 @@ func CreateDeltas(manifest, statedir string, from, to uint32) ([]*File, error) {
+		_ = os.Remove(deltaPath) // Might have returned FULLDL
+		return errors.Wrapf(err, "Failed to create delta for %s -> %s", oldPath, newPath)
 	}
 
-	for _, file := range manifest.Files {
-		if file.Version <= from {
-			continue
-		}
-		if file.DeltaPeer == nil {
+	// Check that the delta actually applies correctly.
+	testPath := filepath.Join(deltaDir, "."+deltaName+".testnewfile")
+	if err := helpers.RunCommandSilent("bspatch", oldPath, testPath, deltaPath); err != nil {
+		return errors.Wrapf(err, "Failed to apply delta %s", deltaPath)
+	}
+	testHash, err := Hashcalc(testPath)
+	if err != nil {
+		return errors.Wrap(err, "Failed to calculate hash for test file created applying delta")
+	}
+	if testHash != delta.To.Hash {
+		return errors.Wrapf(err, "Delta mismatch: %s -> %s via delta: %s", oldPath, newPath, deltaPath)
+	}
+	_ = os.Remove(testPath)
+
+	delta.path = deltaPath
+	return nil
+}
+
+// Delta represents a delta file between two other files. If Error is present, it
+// indicates that the delta couldn't be created.
+type Delta struct {
+	From  *File
+	To    *File
+	Error error
+
+	path string
+}
+
+func findDeltas(c *config, oldManifest, newManifest *Manifest) ([]Delta, error) {
+	// TODO: Pre-allocate with the maximum size possible (wasteful) or trust the reallocs?
+	var result []Delta
+	newIndex := 0
+	oldIndex := 0
+	for newIndex < len(newManifest.Files) && oldIndex < len(oldManifest.Files) {
+		nf := newManifest.Files[newIndex]
+		of := oldManifest.Files[oldIndex]
+
+		if nf.Name < of.Name {
+			// File is not on the old manifest, no chance of delta.
+			newIndex++
 			continue
 		}
-		if file.Type != TypeFile ||
-			file.DeltaPeer.Type != TypeFile {
+		if nf.Name > of.Name {
+			// New file that wasn't present before, no chance of delta.
+			oldIndex++
 			continue
 		}
-
-		deltadir := filepath.Join(c.outputDir, fmt.Sprintf("%d", file.Version), "delta")
-		fromToString := fmt.Sprintf("%d-%d-%s-%s", file.DeltaPeer.Version, file.Version, file.DeltaPeer.Hash, file.Hash)
-
-		deltaFile := filepath.Join(deltadir, fromToString)
-
-		// Only add to list if the delta file does not already exist on the system and has enough size.
-		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
-			fullname := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full", file.Name)
-			file.Info, err = os.Stat(fullname)
+		if nf.Version > of.Version && nf.Present() && of.Present() && nf.Type == TypeFile && !sameFile(nf, of) {
+			// Check file size to decide whether it should have a delta.
+			newPath := filepath.Join(c.imageBase, fmt.Sprint(nf.Version), "full", nf.Name)
+			fi, err := os.Stat(newPath)
 			if err != nil {
 				return nil, errors.Wrapf(err, "error accessing %s to decide whether it can have a delta or not")
 			}
-			if file.Info.Size() < minimumSizeToMakeDeltaInBytes {
-				continue
+			if fi.Size() >= minimumSizeToMakeDeltaInBytes {
+				result = append(result, Delta{From: of, To: nf})
```

Comment:
If we are taking this approach, we don't need DeltaPeer in the File structure.

Whilst I feel that the DeltaPeer doesn't really belong in the File structure, I wonder if this Delta structure is just a little too remote. Really the Delta is a property of two manifests, and we bodge it by saying that files belong to a manifest and the delta information for those files is to the previous manifest.

Reply:
I'll take a middle ground here between the WIP and the current state to make this change less intrusive:

- Keep using DeltaPeer (so zero impact for renaming work you are doing); I'll figure out how to handle avoiding the Version changes in linkedPeers when calculating deltas, as well as resetting the DeltaPeers.
- Make Delta not exported, and use it inside WritePack (so WritePack will call the creation of deltas), this reduces its scope a bit. I could even go as far as say these are not deltas but entries in the delta creation operation. The creation of deltas kind already do 1/3 of that by giving me the failed deltas (as strings and not structured information). The file information is not essential but we use it to be able to pick a fallback.